### PR TITLE
Create rankings for events

### PR DESCRIPTION
**Pull request by birneamstiel**
See #125

### CODE REVIEW

File: app/views/events/ranking.html.erb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    <th><%= t '.lost_count' %></th>
+    <th><%= t '.goals' %></th>
+    <th><%= t '.goals_against' %></th>
+    <th><%= t '.goals_difference' %></th>
+    <th><%= t '.points' %></th>
+  </tr>
+  </thead>
+  <tbody>
+    <% @ranking_entries.each do |entry| %>
+    <tr>
```
Comment (birneamstiel) [inline app/views/events/ranking.html.erb:22]: Whole block could be replaced by something like this:
``` html
<tr>
  <% entry.values do |value| %>
    <td><%= value %></td>
  <% end %>
</tr>
```
Not sure what is more readable @bakoe @jonasnoki

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -83,6 +88,80 @@ def schedule
     @schedule_type = @event.type.downcase!
   end
 
+  # GET /events/1/ranking
+  def ranking
```
Comment (birneamstiel) [inline app/controllers/events_controller.rb:102]: @bakoe @jonasnoki was sagt ihr zu dem jetzigen kontrollfluss? Mir erschien es am lesbarsten, den ranking_entry als parameter durchzureichen und dort die daten einzutragen...
Eine andere denkbare Möglichkeit wäre, recht viel der Logik dem Match zugeben. Dann hätte das Match Methoden wie `#get_ranking_entry_for_home_team` oder so... Was sagt ihr?

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -83,6 +88,80 @@ def schedule
...
+      ranking_entry.rank = index + 1
+    end
+  end
+
+  def parse_matches_data_into_ranking_entry(team, ranking_entry, matches, parse_match_details_for_home_or_away)
+    matches.each do |match|
+      score_home = match.score_home
+      score_away = match.score_away
+      match_has_score = score_home && score_away
+      next unless match_has_score
```
Comment (bakoe) [inline app/controllers/events_controller.rb]: Nice! 👍

File: app/views/events/ranking.html.erb (hunk 1 above)
Comment (bakoe) [reply app/views/events/ranking.html.erb:22]: Since the table's headings (lines 8 to 17 in the same file) cannot be localized dynamically from the Struct's values (or can they?), I would prefer the currently implemented solution of manually listing all of the Struct's properties in `<td>` tags. 

However, if there is an elegant solution of localizing the table's headings dynamically based on the Struct's properties/fields, we could replace both the actual value rendering (lines 23 to 32) **and** the table heading rendering (lines 8 to 17) with something like the snippet you proposed.

File: app/controllers/events_controller.rb [hunk 4]
```diff
@@ -83,6 +88,80 @@ def schedule
...
+    # TODO Either add the possibility to calculate rankings for single-player events
+    # TODO or replace `teams` with the to be introduced EventParticipant superclass
+    # Leaves the Array of RankingEntry Structs empty when no teams participate in the event
+    @event.teams.each do |team|
+      ranking_entry = RankingEntry.new(nil, team.name, 0, 0, 0, 0, 0, 0, 0, 0)
+
+      event_matches = @event.matches
+      # Considers only the team's home matches that belong to the event
+      home_matches_in_event = team.home_matches & event_matches
+      parse_matches_data_into_ranking_entry team, ranking_entry, home_matches_in_event, :parse_match_details_for_home
```
Comment (bakoe) [inline app/controllers/events_controller.rb:113]: We could think about using destructuring (as described e. g. [here](http://tony.pitluga.com/2011/08/08/destructuring-with-ruby.html)) instead of passing the object to be worked on (the `ranking_entry`). The could could look like this:

```ruby
added_match_count, added_won_count, added_draw_count, 
added_lost_count, added_goals, added_goals_against, 
added_points = get_ranking_entry_additions_from_match team, 
                                                      home_matches_in_event, 
                                                      :get_ranking_entry_additions_for_home

ranking_entry.match_count += added_match_count
ranking_entry.won_count += added_won_count
ranking_entry.draw_count += added_draw_count
ranking_entry.lost_count += added_lost_count
ranking_entry.goals += added_goals
ranking_entry.goals_against += added_goals_against
ranking_entry.points += added_points
```

However, while writing this comment I noticed that this approach would result in way more bloated and less readable code. So we should stick with the way of passing the `ranking_entry` object around that you introduced! :+1:

**Review by bakoe — COMMENTED** [review]
@birneamstiel I reflected on your approach regarding the passing of the `ranking_entry` object and the name-based method calling via `send` and came to the conclusion that the approach you chose seems very reasonable to me! I would not change the implementation of the EventsController's `ranking` method and its helper methods. 👍

File: app/controllers/events_controller.rb (hunk 2 above)
Comment (bakoe) [reply app/controllers/events_controller.rb:102]: As mentioned in my PR feedback, I approve the current control flow. I would prefer to not move the helper methods into the `Match` class since this would require that one more component knows about the RankingEntry Struct and knows how to handle it properly. 

Maybe a future user story will require to re-use the RankingEntry-specific methods of the EventsController in some other controller – to my mind, this would be a good time to factor out the methods into the `Match` class. But as long as this encapsulation is not needed and the RankingEntry Structs get only used within the `Event` scope, I would leave the methods were they are currently.

File: app/controllers/events_controller.rb [hunk 5]
```diff
@@ -93,6 +100,80 @@ def schedule
...
+    ranking_entry.goals_against += score_away
+    ranking_entry.points += match.points_home
+  end
+
+  def parse_match_details_for_away(ranking_entry, score_home, score_away, match)
+    ranking_entry.goals += score_away
+    ranking_entry.goals_against += score_home
+    ranking_entry.points += match.points_away
+  end
+
```
Comment (lheilmann) [inline app/controllers/events_controller.rb:175]: Great control flow and encapsulation of functionality in the methods above. :+1:

File: app/views/events/ranking.html.erb (hunk 1 above)
Comment (lheilmann) [reply app/views/events/ranking.html.erb:22]: I agree with @bakoe. In my opinion it is more readable, when both parts are either composed to a snippet or decomposed.

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -98,6 +98,10 @@
             event_overview_path(@event),
             class: 'btn btn-default' %>
 
+<%= link_to t('.to_ranking'),
```
Comment (lheilmann) [inline app/views/events/show.html.erb:105]: Please use consistent translation function call - either with or without brackets. See line 8 to 17 in same file.

File: spec/controllers/events_controller_spec.rb [hunk 7]
```diff
@@ -362,6 +378,120 @@
...
+    end
+    describe 'when teams have joined the event' do
+      describe 'when no match has been played yet' do
+        it 'should calculate a lexicographically sorted zero-filled ranking' do
+          event = League.create! valid_league_attributes
+          team1 = FactoryBot.create(:team)
+          team2 = FactoryBot.create(:team)
+          event.add_team(team1)
+          event.add_team(team2)
+          event.generate_schedule
```
Comment (lheilmann) [inline spec/controllers/events_controller_spec.rb]: Why do you need to generate a schedule in `when no match has been played yet` but not in the following test cases in `when at least two matches have been played already`?

File: spec/controllers/events_controller_spec.rb [hunk 8]
```diff
@@ -362,6 +378,120 @@
...
+        it 'should sum up the own goals of a participant correctly' do
+          expect(@ranking_entries.first.goals).to eq(20)
+        end
+        it 'should sum up the goals for the other side of a participant correctly' do
+          expect(@ranking_entries.first.goals_against).to eq(11)
+        end
+        it 'should calculate the number of points of a participant correctly' do
+          expect(@ranking_entries.first.points).to eq(4)
+        end
+        it 'should calculate the rank of two participants with the same points based on the number of goals correctly' do
```
Comment (lheilmann) [inline spec/controllers/events_controller_spec.rb:460]: Cool! As far as I have seen you covered all possible orders of RankingEntries here! :+1:

File: spec/controllers/events_controller_spec.rb [hunk 9]
```diff
@@ -362,6 +378,120 @@
...
+          expect(@ranking_entries.first.name).to eq(@team1.name)
+          expect(@ranking_entries[1].name).to eq(@team3.name)
+
+        end
+      end
+    end
+    describe 'when users have joined the event' do
+      # TODO Factor out and reference tests for events with joined teams from above
```
Comment (lheilmann) [inline spec/controllers/events_controller_spec.rb]: I do not understand your `TODO` comments here. Are these test case moved to a different location or not implemented yet?

File: spec/views/events/show.html.erb_spec.rb [hunk 10]
```diff
@@ -50,6 +50,11 @@
       render
       expect(rendered).to_not have_selector(:link_or_button, t('helpers.links.destroy'))
     end
+
+    it 'has a ranking button' do
```
Comment (lheilmann) [inline spec/views/events/show.html.erb_spec.rb:54]: Nice!

File: spec/controllers/events_controller_spec.rb [hunk 11]
```diff
@@ -53,6 +53,22 @@
     FactoryBot.build(:league, name: nil).attributes
   }
 
+  let(:valid_small_league_attributes) {
+    FactoryBot.build(:league, owner: @user, max_teams: 2).attributes
+  }
+
+  let(:valid_team1_attributes) {
+    FactoryBot.attributes_for(:team)
```
Comment (lheilmann) [inline spec/controllers/events_controller_spec.rb]: I assume the method `attributes_for` also creates an instance of Team by FactoryBot and then returns the attributes? Then `valid_team1_attributes`, `valid_team2_attributes` and `valid_team3_attributes` are instantiated with different values, which is probably the desired behaviour? I am asking because I have only used this method for returning attributes of an already created instance earlier.

**Review by lheilmann — CHANGES_REQUESTED** [review]
Looks very good to me all in all! Especially control flow in EventsController is pretty cool and well-considered! :+1: I have only reviewed the code so far, because acceptance tests are not possible right now due to bugs in the GUI. Will do later, as soon as GUI is working properly again. Not our job...

**Comment by lheilmann** [comment]
No need to fix all codeclimate issues in my opinion. Especially the issues targeting the control flow are not quite reasonable I think. So you can resolve some of them, as long as you have the same opinion. What about the `TODO`'s and `FIXME`'s?

File: app/controllers/events_controller.rb [hunk 12]
```diff
@@ -93,6 +100,80 @@ def schedule
...
+      next unless match_has_score
+
+      ranking_entry.match_count += 1
+      parse_match_result_into_ranking_entry team, match, ranking_entry
+
+      send(parse_match_details_for_home_or_away, ranking_entry, score_home, score_away, match)
+    end
+  end
+
+  def parse_match_result_into_ranking_entry(team, match, ranking_entry)
```
Comment (Askir) [inline app/controllers/events_controller.rb:152]: All these methods don't really belong to the controller. Try to keep as much logic as possible outside of the controller. Methods that aren't actions need a good reason to even exist. And in this case it is very model specific. I would suggest encapsulation these methods into the Rankinglist model or even better into seperate plain old ruby objects (POROs) that handle all the neseccary logic.
The ranking function should do much more than setting `@ranking_entries = @event.ranking_entries`.

Comment (Askir) [reply app/controllers/events_controller.rb:152]: Here is a somewhat related article on the matter of POROs: http://blog.steveklabnik.com/posts/2011-09-06-the-secret-to-rails-oo-design

**Comment by lheilmann** [comment]
Started with acceptance tests after design issues are fixed. Please see screenshot. Seems to me, that this is not the expected behaviour? Will continue with acceptance tests tomorrow in the morning.
screen shot 2018-01-17 at 00 12 37

**Comment by bakoe** [comment]
@lheilmann I looked into the issue you documented and found out that it was probably caused by a bug/counter-intuitiveness in the “Spielplan” view:

The problem is: The “Spielplan” view currently makes the user think that the points entered in the left text field are the `points_home` (since the left team is `team_home` and the left side of the goals is `score_home`). **But:** The points entered in the left text field are actually `points_away` and the points entered in the right text field are actually `points_home`. See the following screenshot for explanation:

bildschirmfoto 2018-01-17 um 00 46 49 kopie

The corresponding `Match` object looks as follows, proving the documented bug/counter-intuitiveness:

bildschirmfoto 2018-01-17 um 00 47 37

As you can see, I as a user would expect the (left) `score_home` to be 3 and the (right) `score_away` to be 0. However, the opposite is the case.

Because of this bug, the “Spielplan” results from my screenshot result in the following ranking table which looks pretty similarly wrong to the one you posted @lheilmann:

bildschirmfoto 2018-01-17 um 00 47 46

However, the ranking table works actually as expected since the ranking entries (more precise: especially the amount of won, drawn and lost games as well as the amount of points) are **calculated based on the points** and _not based on the goals_. This is caused by the current match architecture of having both goals and (redundant) points – we decided to calculate the ranking entries based on the points since the points should theoretically get calculated from the goals automatically (see #100 with still open PR #292). The goals only get used for calculating the rank when two teams have the same amount of points.

TL;DR: Problem not caused by EventsController's `ranking` method but by a bug/counter-intuitiveness in the “Spielplan” view's text field ordering which causes the user to enter wrong points for matches. When #292 is merged (and thus #100 is closed), the problem should not occur anymore since the user should not have to enter points manually anymore.

**Comment by lheilmann** [comment]
Thanks for the detailed explanation @bakoe! I would like to review this PR (acceptance tests) together with PO @svitlanakalancha in our meeting today. That's just because I think you need to know some tricky stuff or bugs, caused by other teams, to reproduce the desired behaviour. We then have maximum half an hour to review this issue all together. 

I approve this PR, because you covered all use case with test cases, so we can tell `ranking` is pretty much doing what it should be doing.

Please have a look at the codeclimate issues, so that we are able to merge this PR into `dev` before our meeting at 1 PM. 

Hang loose :+1:

**Review by lheilmann — APPROVED** [review]

File: app/controllers/events_controller.rb (hunk 12 above)
Comment (birneamstiel) [reply app/controllers/events_controller.rb:152]: Thanks! Refactoring will follow up:  https://github.com/hpi-swt2/sport-portal/issues/367
